### PR TITLE
Feature/1 basic login (and some other things)

### PR DESCRIPTION
**Pull request by Dragon1320**
Basic setup required to do everything (db, helpers, etc.).

Also the following routes:
- /2fa/secret
- /@me/2fa
- /@me/2fa
- /@me/2fa/codes
- /@me/2fa/codes/refresh
- /register
- /verify
- /verify/resend
- /login
- /recover
- /recover/verify

What ill change soon:tm::
- group all oauth login stuff under a single /authenticate route
- fix mfa code generation and check them during login

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -25,6 +25,11 @@ module.exports = {
     "no-console": "off",
     "arrow-parens": "off",
     "max-len": ["error", 150],
+    "@typescript-eslint/interface-name-prefix": "off",
+    // doesnt work with typescript stuff
+    "no-undef": "off",
+    "no-bitwise": "off",
+    "no-plusplus": "off",
```
Comment (MattheousDT) [inline .eslintrc.js]: Remove: https://github.com/airbnb/javascript#variables--unary-increment-decrement

File: package.json [hunk 2]
```diff
@@ -42,7 +49,20 @@
     "typescript": "^3.8.3"
   },
   "dependencies": {
+    "@typegoose/typegoose": "^6.4.0",
+    "@types/uuid": "^7.0.2",
```
Comment (MattheousDT) [inline package.json]: Add to dev dependencies

File: src/db/schemas/base.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
+/* eslint-disable max-classes-per-file */
```
Comment (MattheousDT) [inline src/db/schemas/base.ts]: You only have one class m8

File: src/db/services/base.ts [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        deletedCount,
+      };
+    } catch (error) {
+      throw new Error(`error running BaseService.delete ( searchInfo: ${searchInfo} ): ${error}`);
+    }
+  }
+}
+
+export {
+  // eslint-disable-next-line import/prefer-default-export
```
Comment (MattheousDT) [inline src/db/services/base.ts]: Don't need that bruh

File: src/db/services/verification.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
+import { Document } from "mongoose";
+
+import { BaseService } from "./base";
+import { Verification } from "../schemas";
+// import { IVerification } from "../types";
```
Comment (MattheousDT) [inline src/db/services/verification.ts]: does this need to be commented out pt 1 the miracle and the sleeper

File: src/db/services/verification.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+import { VerificationModel } from "../models";
+
+// interface IVerificationService extends IBaseService<Verification> {}
```
Comment (MattheousDT) [inline src/db/services/verification.ts]: Does this need to be commented out pt 2 scenes from a memory

File: src/server/helpers/email/types/email.ts [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+  protected subject!: string;
+
+  constructor(transportOptions: IEmailTransportOptions, templateName: string) {
+    this.transporter = createTransport(transportOptions);
+
+    this.displayName = DEFAULT_DISPLAY_NAMME;
+    this.display = DEFAULT_DISPLAY;
+
+    this.template = fs.readFileSync(path.join(__dirname, "..", "..", "..", "..", "..", "templates", templateName)).toString();
```
Comment (MattheousDT) [inline src/server/helpers/email/types/email.ts]: try use node path or some shit idk

File: src/server/helpers/oauth.ts [hunk 8]
```diff
@@ -0,0 +1,186 @@
+/* eslint-disable @typescript-eslint/camelcase */
+import querystring, { ParsedUrlQueryInput } from "querystring";
+
+import fetch from "node-fetch";
+
+import { OauthServiceType } from "../../db";
+
+// typescript is just a pile of wank innit
```
Comment (MattheousDT) [inline src/server/helpers/oauth.ts:8]: dont you slag my bae like that >:(

File: src/server/routes/v1/index.ts [hunk 9]
```diff
@@ -1,110 +1,1213 @@
...
+        reply.code(400);
+
+        return {
+          error: "oauth service type or access token not specified",
+        };
+      }
+
+      // check this as exchangeToken expects an enum value, which cant be checked at compile
+      // time cos user input and such, passing an incorrect value wont break anything, it just
+      // wont throw a pretty looking error like the one below ^^
```
Comment (MattheousDT) [inline src/server/routes/v1/index.ts:657]: either you're retarded or missing a mouth

File: src/server/routes/v1/index.ts [hunk 10]
```diff
@@ -1,110 +1,1213 @@
...
```
Comment (MattheousDT) [inline src/server/routes/v1/index.ts:999]: ok, this is just getting offensive

**Review by MattheousDT — CHANGES_REQUESTED** [review]

File: .eslintrc.js (hunk 1 above)
Comment (Dragon1320) [reply .eslintrc.js]: done

File: package.json (hunk 2 above)
Comment (Dragon1320) [reply package.json]: done

File: src/db/schemas/base.ts (hunk 3 above)
Comment (Dragon1320) [reply src/db/schemas/base.ts]: done

File: src/db/services/base.ts (hunk 4 above)
Comment (Dragon1320) [reply src/db/services/base.ts]: done

File: src/db/services/verification.ts (hunk 5 above)
Comment (Dragon1320) [reply src/db/services/verification.ts]: done

File: src/db/services/verification.ts (hunk 6 above)
Comment (Dragon1320) [reply src/db/services/verification.ts]: done

File: src/server/helpers/email/types/email.ts (hunk 7 above)
Comment (Dragon1320) [reply src/server/helpers/email/types/email.ts]: done

File: src/server/helpers/oauth.ts (hunk 8 above)
Comment (Dragon1320) [reply src/server/helpers/oauth.ts:8]: fuk u

File: src/server/routes/v1/index.ts (hunk 9 above)
Comment (Dragon1320) [reply src/server/routes/v1/index.ts:657]: image

File: src/server/routes/v1/index.ts (hunk 10 above)
Comment (Dragon1320) [reply src/server/routes/v1/index.ts:999]: no u

**Review by MattheousDT — APPROVED** [review]

**Pull request merged**